go-controller/pkg/node/udn_isolation.go

```diff
@@ -0,0 +1,459 @@
+func (m *UDNHostIsolationManager) runKubeletRestartTracker(ctx context.Context) (err error) {
+	conn, err := dbus.NewSystemdConnectionContext(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to connect to systemd: %w", err)
+	}
+	defer func() {
+		if err != nil {
+			conn.Close()
+		}
+	}()
+
+	err = conn.Subscribe()
+	if err != nil {
+		return fmt.Errorf("failed to subscribe to systemd events: %w", err)
+	}
+	// interval is important here as we need to catch the restart state, before it is running again
+	events, errChan := conn.SubscribeUnitsCustom(50*time.Millisecond, 0, func(u1, u2 *dbus.UnitStatus) bool { return *u1 != *u2 },
+		func(s string) bool {
+			return s != "kubelet.service"
+		})
+	// run until context is cancelled
+	go func() {
+		waitingForActive := false
+		for {
+			select {
+			case <-ctx.Done():
+				conn.Close()
+				return
+			case event := <-events:
+				for _, status := range event {
+					if status.ActiveState != "active" {
+						waitingForActive = true
+					} else if waitingForActive {
+						klog.Infof("Kubelet was restarted, re-applying UDN host isolation")
+						err = m.updateKubeletCgroup()
+						if err != nil {
+							klog.Errorf("Failed to re-apply UDN host isolation: %v", err)
+						} else {
+							waitingForActive = false
+						}
+					}
+				}
+			case err := <-errChan:
+				klog.Errorf("Systemd listener error: %v", err)
+			}
+		}
+	}()
+	return nil
+}
+
+func (m *UDNHostIsolationManager) podInitialSync() error {
+	pods, err := m.podLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list pods: %v", err)
+	}
+	v4PodsIPs := map[string]sets.Set[string]{}
+	v6PodsIPs := map[string]sets.Set[string]{}
+	for _, pod := range pods {
+		// only add pods with primary UDN
+		primaryUDN, err := m.isPodPrimaryUDN(pod)
+		if err != nil {
+			return fmt.Errorf("failed to check if pod %s in namespace %s is in primary UDN: %w", pod.Name, pod.Namespace, err)
+		}
+		if !primaryUDN {
+			continue
+		}
+
+		podIPs, err := util.DefaultNetworkPodIPs(pod)
+		if err != nil {
+			return fmt.Errorf("failed to get pod IPs for pod %s in namespace %s: %v", pod.Name, pod.Namespace, err)
+		}
+		if podIPs == nil {
+			continue
+		}
+		newV4IPs, newV6IPs := splitIPsPerFamily(podIPs)
+		podKey, err := cache.MetaNamespaceKeyFunc(pod)
+		if err != nil {
+			klog.Warningf("UDNHostIsolationManager failed to get key for pod %s in namespace %s: %v", pod.Name, pod.Namespace, err)
+			continue
+		}
+		v4PodsIPs[podKey] = newV4IPs
+		v6PodsIPs[podKey] = newV6IPs
+	}
+	if err = m.udnPodIPsv4.fullSync(m.nft, v4PodsIPs); err != nil {
+		return err
+	}
+	if err = m.udnPodIPsv6.fullSync(m.nft, v6PodsIPs); err != nil {
+		return err
+	}
+	return nil
+}
+
+func podNeedsUpdate(oldObj, newObj *v1.Pod) bool {
+	if oldObj == nil || newObj == nil {
+		return true
+	}
+	// react to pod IP changes
+	return !reflect.DeepEqual(oldObj.Status, newObj.Status) || !reflect.DeepEqual(oldObj.Annotations, newObj.Annotations)
```

Comment:
nit: might be a little better to only scope the deep equal on the specific k8s.ovn.org pod annotation